### DISCUSSION TITLE
Drop the index of the foreign using method `dropForeign`

### DISCUSSION THREAD

**Discussion by serious-angel** [laravel/framework#53155] [Ideas]
Dear Community and Developers,

Thank you very much for the ineffably marvelous project! :sparkles: 

May I ask regarding the method [`dropForeign`](<https://github.com/laravel/framework/blob/1d7e000619be1ffe24e4fee93f11a1fcf5af841e/src/Illuminate/Database/Schema/Blueprint.php#L503>)? Currently, when we create a new foreign, Laravel creates a corresponding `index` for the constraint. For example, let's check a quick case.

# Case

## Initial state of table "test'

```php
dump(DB::select('SHOW CREATE TABLE `test`'));
```

```php
array:1 [
  0 => {#7128
    +"Table": "test"
    +"Create Table": """
      CREATE TABLE `test` (\n
        `id` int unsigned NOT NULL AUTO_INCREMENT,\n
        `user_id` int NOT NULL,\n
        `notification` int unsigned DEFAULT NULL,\n
        `comment` text COLLATE utf8_unicode_ci,\n
        `created_at` timestamp NULL DEFAULT NULL,\n
        `updated_at` timestamp NULL DEFAULT NULL,\n
        `deleted_at` timestamp NULL DEFAULT NULL,\n
        PRIMARY KEY (`id`),\n
        KEY `user_id_IDX` (`user_id`),\n
        KEY `test_created_at_index` (`created_at`),\n
      ) ENGINE=InnoDB AUTO_INCREMENT=524 DEFAULT CHARSET=utf8mb3 COLLATE=utf8_unicode_ci
      """
  }
]
```

## Create a foreign for table "notifications"

```php
Schema::table('test', function (\Illuminate\Database\Schema\Blueprint \$table) {
  \$table->foreign('notification')
    ->references('id')->on('notifications')
    ->restrictOnUpdate()
    ->restrictOnDelete();
});

dump(DB::select('SHOW CREATE TABLE `test`'));
```

```php
array:1 [
  0 => {#7128
    +"Table": "test"
    +"Create Table": """
      CREATE TABLE `test` (\n
        `id` int unsigned NOT NULL AUTO_INCREMENT,\n
        `user_id` int NOT NULL,\n
        `notification` int unsigned DEFAULT NULL,\n
        `comment` text COLLATE utf8_unicode_ci,\n
        `created_at` timestamp NULL DEFAULT NULL,\n
        `updated_at` timestamp NULL DEFAULT NULL,\n
        `deleted_at` timestamp NULL DEFAULT NULL,\n
        PRIMARY KEY (`id`),\n
        KEY `user_id_IDX` (`user_id`),\n
        KEY `test_created_at_index` (`created_at`),\n
        KEY `test_notification_foreign` (`notification`),\n
        CONSTRAINT `test_notification_foreign` FOREIGN KEY (`notification`) REFERENCES `notifications` (`id`) ON DELETE RESTRICT ON UPDATE RESTRICT,\n
      ) ENGINE=InnoDB AUTO_INCREMENT=524 DEFAULT CHARSET=utf8mb3 COLLATE=utf8_unicode_ci
      """
  }
]
```

As we see, alongside the foreign constraint `test_notification_foreign` there an additional index `test_notification_foreign` has just been automatically created. The rationale behind the creation is fairly clear, since the performance difference with an without such an index may be crucial.

## Drop the created foreign

```php
Schema::table('test', function (\Illuminate\Database\Schema\Blueprint \$table) {
  \$table->dropForeign(['notification']);
});

dump(DB::select('SHOW CREATE TABLE `test`'));
```

```php
array:1 [
  0 => {#7128
    +"Table": "test"
    +"Create Table": """
      CREATE TABLE `test` (\n
        `id` int unsigned NOT NULL AUTO_INCREMENT,\n
        `user_id` int NOT NULL,\n
        `notification` int unsigned DEFAULT NULL,\n
        `comment` text COLLATE utf8_unicode_ci,\n
        `created_at` timestamp NULL DEFAULT NULL,\n
        `updated_at` timestamp NULL DEFAULT NULL,\n
        `deleted_at` timestamp NULL DEFAULT NULL,\n
        PRIMARY KEY (`id`),\n
        KEY `user_id_IDX` (`user_id`),\n
        KEY `test_created_at_index` (`created_at`),\n
        KEY `test_notification_foreign` (`notification`),\n
      ) ENGINE=InnoDB AUTO_INCREMENT=524 DEFAULT CHARSET=utf8mb3 COLLATE=utf8_unicode_ci
      """
  }
]
```

As we see, the foreign constraint `test_notification_foreign` got removed but the index `test_notification_foreign`. This may require an additional explicit command to be executed after the `dropForeign`:

```php
\$table->dropForeign(['notification']);
\$table->dropIndex('test_notification_foreign');
```

# Question

1. Is it possible to have an additional parameter in the method `dropForeign` (e.g. `index` of type `string|bool` where `string` would define an explicit name, and `bool` - an self-generated by the framework)?
2. Is the rationale of leaving the automatically created index the performance or the unknown name of the key previously created, considering the following condition?:

https://github.com/laravel/framework/blob/1d7e000619be1ffe24e4fee93f11a1fcf5af841e/src/Illuminate/Database/Schema/Blueprint.php#L1633-L1640

Best and kind regards

# Related

\- [Dropping column with foreign key Laravel error](<https://stackoverflow.com/a/78964705/5113030>)

**Comment by hafezdivandari** [comment]
> Currently, when we create a new foreign, Laravel creates a corresponding index for the constraint. For example, let's check a quick case.

This is not true, Laravel doesn't create an index for the foreign constraint, MySQL does! [MySQL requires indexes on foreign keys](https://dev.mysql.com/doc/refman/8.4/en/create-table-foreign-keys.html#:~:text=MySQL%20requires%20indexes,as%20described%20previously.) and referenced keys, and an index is created on the referencing table automatically if it does not exist.

> Is it possible to have an additional parameter in the method dropForeign (e.g. index of type string|bool where string would define an explicit name, and bool - an self-generated by the framework)?

Not a good idea, because it's MySQL specific, for example PostgreSQL doesn't create an index / doesn't need an index on foreign constraint.

> Is the rationale of leaving the automatically created index the performance or the unknown name of the key previously created, considering the following condition?:

Didn't understand your question!

## What we can improve here?

We can improve `dropForeign` and `dropIndex` methods. Currently when passing column name(s) as array to these function, Laravel will use the conventional name to drop foreign / index. But it's much better to use this column array to find foreign / indexes names (using `Schema::getIndexes()` and `Schema::getForeigns()` methods) and remove them. then you can simply do:

```php
$table->dropForeign(['notification']); // drop any foreign key constraint of `['notification']` column not just `test_notification_foreign`
$table->dropIndex(['notification']); // drops any indexes of `['notification']` column not just `test_notification_index`
```
